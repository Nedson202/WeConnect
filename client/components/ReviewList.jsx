import 'rc-pagination/assets/index.css';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import Pagination from 'rc-pagination';
import StarRatingComponent from 'react-star-rating-component';
import avatar from '../images/user-avatar.png';
import ReviewList from './business-profile/ReviewList';

/**
 * @class BusinessList
 * @extends {Component}
 */
class ReviewsList extends Component {
  /**
   * @description Creates an instance of BusinessList.
   * @param {object} props
   * @memberof Profile
   */
  constructor(props) {
    super(props);

    this.state = {
      current: 1
    };

    this.onChange = this.onChange.bind(this);
    this.onReviewDelete = this.onReviewDelete.bind(this);
  }

  /**
   * @description Fetch reviews and business
   * 
   * @returns {undefined}
   * 
   * @memberof BusinessProfile
   */

  /**
   * @description Toggler for pagination
   *
   * @param {any} page
   *
   * @returns {undefined}
   *
   * @memberof BusinessList
   */
  onChange(page) {
    const genPage = `page=${page}`;

    this.setState({
      current: page,
    });

    this.props.fetchReviews(this.props.params.id, genPage);
  }

  /**
   * @description Delere business
   *
   * @param {any} businessId
   *
   * @param {any} reviewId
   *
   * @returns {undefined}
   *
   * @memberof ReviewList
   */
  onReviewDelete(businessId, reviewId) {
    return () => {
      this.props.deleteReview(businessId, reviewId).then(() => {
        this.props.fetchBusinessById(businessId);
      });
    };
  }

  /**
   * @description Renders the component to the dom
   *
   * @returns {object} JSX object
   *
   * @memberof ReviewList
   */
  render() {
    const {
      reviews, params, user, paginate
    } = this.props;
    const { username } = user;

    const paginateAll = (
      <div>{ reviews.length > 0 ?
        <div className="d-flex justify-content-center">
          { paginate.limit === undefined ? null :
          <Pagination
            onChange={this.onChange}
            current={this.state.current}
            total={!paginate ? null : paginate.totalReviews}
            defaultCurrent={this.state.current}
            defaultPageSize={!paginate ? null : paginate.limit}
          />
          }
        </div> : null}
      </div>
    );

    // console.log(paginate);


    return (
      <div>
        {reviews.map(review => (
          <div key={review.id}>
            { !review.reviewer.image ? <img src={avatar} className="rounded-circle user-avatar" alt="Avatar" width="50" height="50" />
            : <img src={review.reviewer.image} className="rounded-circle user-avatar" alt="Avatar" width="50" height="50" />
          }
            <div className="review-body">
              <span className="textcase reviewer">{review.reviewer.username} </span>
              {
              (username !== review.reviewer.username && username !== 'admin') ? null
              : (username === review.reviewer.username || username === 'admin') ?
                <button onClick={this.onReviewDelete(params.id, review.id)} className="close"><span>&times;</span></button> : null
            }
              <p>{review.message}</p>
              <StarRatingComponent
                name="rate2"
                editing={false}
                starCount={5}
                value={review.rating}
                starColor="#ffd700"
              />
              <p
                className="mb-0 small font-weight-medium
              mb-1 text-muted lts-2px"
              >
                {moment(review.createdAt).format('Do MMMM YYYY HH:mm')}
              </p>
            </div>
          </div>))}

        { !paginate ? null
          : paginate.totalReviews <= paginate.limit ? null : paginateAll}
      </div>
    );
  }
}

ReviewList.propTypes = {
  fetchReviews: PropTypes.func.isRequired,
  deleteReview: PropTypes.func.isRequired,
};

export default ReviewsList;
