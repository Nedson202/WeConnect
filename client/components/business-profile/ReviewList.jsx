import 'rc-pagination/assets/index.css';
import React from 'react';
import PropTypes from 'prop-types';
import Pagination from 'rc-pagination';
import Review from './Review';
/**
 * @class BusinessList
 * @extends {Component}
 */
const ReviewsList = ({
  reviews,
  params,
  user,
  paginate,
  deleteReview,
  fetchBusinessById,
  reviewUpdateRequest,
  fetchReviews,
  onChange,
  onStarClick,
  state,
  edit,
  cancelEdit,
  deleteStatus,
  onReviewDelete,
  onReviewPageChange,
  onReviewUpdate,
}) => {
  const { username } = user;
  const { totalReviews, limit } = paginate;

  const paginateAll = (
    <div>{ reviews.length > 0 ?
      <div className="d-flex justify-content-center">
        { paginate && totalReviews === undefined ? null :
        <Pagination
          onChange={onReviewPageChange}
          current={state.current}
          total={!paginate ? null : totalReviews}
          defaultCurrent={state.current}
          defaultPageSize={!paginate ? null : limit}
        />
        }
      </div> : null}
    </div>
  );

  return (
    <div>
      {reviews.map(review =>
        (<Review
          review={review}
          key={review.id}
          params={params}
          username={username}
          deleteReview={deleteReview}
          fetchBusinessById={fetchBusinessById}
          fetchReviews={fetchReviews}
          onStarClick={onStarClick}
          onChange={onChange}
          state={state}
          reviewUpdateRequest={reviewUpdateRequest}
          edit={edit}
          cancelEdit={cancelEdit}
          deleteStatus={deleteStatus}
          onReviewDelete={onReviewDelete}
          onReviewUpdate={onReviewUpdate}
        />))
      }
      <div>
        {
          !paginate ? null :
          <span>
            {totalReviews <= limit ? null : paginateAll}
          </span>
        }
      </div>
    </div>
  );
};
ReviewsList.propTypes = {
  fetchReviews: PropTypes.func.isRequired,
  deleteReview: PropTypes.func.isRequired,
  fetchBusinessById: PropTypes.func.isRequired,
  reviewUpdateRequest: PropTypes.func.isRequired,
  onReviewUpdate: PropTypes.func.isRequired,
  onReviewPageChange: PropTypes.func.isRequired,
  paginate: PropTypes.object,
  state: PropTypes.object.isRequired,
  reviews: PropTypes.array.isRequired,
  params: PropTypes.object.isRequired,
  onStarClick: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  onReviewDelete: PropTypes.func.isRequired,
  deleteStatus: PropTypes.func.isRequired,
  edit: PropTypes.func.isRequired,
  cancelEdit: PropTypes.func.isRequired,
};

ReviewsList.defaultProps = {
  paginate: {}
};

export default ReviewsList;
