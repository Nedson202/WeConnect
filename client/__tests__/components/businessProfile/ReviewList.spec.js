import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';
import ReviewList from '../../../components/business-profile/ReviewList';
import businessData from '../../__mockData__/business';
import userData from '../../__mockData__/userData';

const { business, reviews, review, paginationResult} = businessData;
const { userUpdateData } = userData;
configure({ adapter: new Adapter() });

let props;
const setup = () => {
  props = {
    onChange: jest.fn(() => Promise.resolve()),
    onStarClick: jest.fn(() => Promise.resolve()),
    onSubmit: jest.fn(() => Promise.resolve()),
    fetchReviews: jest.fn(() => Promise.resolve()),
    deleteReview: jest.fn(() => Promise.resolve()),
    deleteStatus: jest.fn(() => Promise.resolve()),
    onReviewDelete: jest.fn(() => Promise.resolve()),
    onReviewUpdate: jest.fn(() => Promise.resolve()),
    onReviewPageChange: jest.fn(() => Promise.resolve()),
    fetchBusinessById: jest.fn(() => Promise.resolve()),
    reviewUpdateRequest: jest.fn(() => Promise.resolve()),
    edit: jest.fn(() => Promise.resolve()),
    cancelEdit: jest.fn(() => Promise.resolve()),
    edit: jest.fn(() => Promise.resolve()),
    state: review,
    user: userUpdateData,
    params: {},
    reviews,
    paginate: paginationResult
  };

  return shallow(<ReviewList {...props} />);
};


describe('Component: ReviewList', () => {
  beforeEach(() => {
    global.window = {
      window: () => {},
    }
  })

  it('should render', () => {
    const wrapper = setup();
    expect(wrapper.find('div').length).toBe(4);
    expect(wrapper.find('span').length).toBe(1);
    expect(wrapper.find('Pagination').length).toBe(1);
    expect(wrapper.find('Pagination').props().defaultPageSize).toBe(6);
    expect(wrapper.length).toBe(1);
    wrapper.setProps({ paginate: {}, reviews: [] });
    expect(wrapper.find('span').length).toBe(1);
    expect(wrapper.find('div').length).toBe(3);
    expect(wrapper.find('Pagination').length).toBe(0);
    wrapper.setProps({ paginate: {
      totalReviews: 3,
      limit: 5
    }, reviews: [] });
    expect(wrapper.find('div').length).toBe(2);
    wrapper.setProps({ paginate: {
      totalReviews: undefined,
      limit: 5
    }, reviews: [] });
    expect(wrapper.find('Pagination').length).toBe(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});