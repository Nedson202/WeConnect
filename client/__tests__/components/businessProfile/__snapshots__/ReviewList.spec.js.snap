// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ReviewList should render 1`] = `
<div>
  <div>
    <span>
      <div />
    </span>
  </div>
</div>
`;
